apps/rush-lib/src/logic/pnpm/PnpmProjectShrinkwrapFile.ts

```diff
@@ -0,0 +1,238 @@
+        dependencies: { ...importer.dependencies, ...importer.devDependencies },
+        optionalDependencies: { ...importer.optionalDependencies }
+      });
+    }
+
+    return projectShrinkwrapMap;
+  }
+
+  protected generateLegacyProjectShrinkwrapMap(): Map<string, string> {
+    const tempProjectDependencyKey: string | undefined = this.shrinkwrapFile.getTempProjectDependencyKey(
+      this.project.tempProjectName
+    );
+    if (!tempProjectDependencyKey) {
+      throw new Error(`Cannot get dependency key for temp project: ${this.project.tempProjectName}`);
+    }
+    const parentShrinkwrapEntry: IPnpmShrinkwrapDependencyYaml = this.shrinkwrapFile.getShrinkwrapEntryFromTempProjectDependencyKey(
+      tempProjectDependencyKey
+    )!;
+
+    // Only select the shrinkwrap dependencies that are non-local since we already handle local
+    // project changes
+    const externalDependencies: [string, string][] = [
+      ...Object.entries(parentShrinkwrapEntry.dependencies || {}),
+      ...Object.entries(parentShrinkwrapEntry.optionalDependencies || {})
+    ].filter((d) => d[0].indexOf('@rush-temp/') === -1);
+
+    const projectShrinkwrapMap: Map<string, string> = new Map();
+    for (const [name, version] of externalDependencies) {
+      this._addDependencyRecursive(projectShrinkwrapMap, name, version, parentShrinkwrapEntry);
+    }
+
+    // Since peer dependencies within on external packages may be hoisted up to the top-level package,
+    // we need to resolve and add these dependencies directly
+    this._resolveAndAddPeerDependencies(projectShrinkwrapMap, parentShrinkwrapEntry);
+
+    return projectShrinkwrapMap;
+  }
+
+  private _addDependencyRecursive(
+    projectShrinkwrapMap: Map<string, string>,
+    name: string,
+    version: string,
+    parentShrinkwrapEntry: IPnpmShrinkwrapDependencyYaml,
+    throwIfShrinkwrapEntryMissing: boolean = true
+  ): void {
+    const shrinkwrapEntry: IPnpmShrinkwrapDependencyYaml | undefined = this.shrinkwrapFile.getShrinkwrapEntry(
+      name,
+      version
+    );
+
+    if (!shrinkwrapEntry) {
+      if (throwIfShrinkwrapEntryMissing) {
+        throw new InternalError(`Unable to find dependency ${name} with version ${version} in shrinkwrap.`);
+      }
+      return;
+    }
+
+    const specifier: string = `${name}@${version}`;
+    let integrity: string | undefined = shrinkwrapEntry?.resolution?.integrity;
+    if (!integrity) {
+      // git dependency specifiers do not have an integrity entry. Instead, they specify the tarball field.
+      // So instead, we will hash the contents of the dependency entry and use that as the integrity hash.
+      // Ex:
+      // github.com/chfritz/node-xmlrpc/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38:
+      //   ...
+      //   resolution:
+      //     tarball: 'https://codeload.github.com/chfritz/node-xmlrpc/tar.gz/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38'
+      const sha256Digest: string = crypto
```

Comment:
You should see how often this gets called.

Reply:
Very rarely, thankfully. Only for packages w/o integrity, which only happens in rare situations (like github-sourced packages)